Panic in fs walk

in the test suite:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x46ce6fa]

goroutine 104 [running]:
github.com/syncthing/syncthing/lib/scanner.(*walker).createFSWalkFn.func1(0xc4208000e0, 0x14, 0x0, 0x0, 0x49bac80, 0xc4201380c0, 0xc4201380c0, 0x0)
	/Users/jb/src/github.com/syncthing/syncthing/lib/scanner/walk.go:268 +0x2aa
github.com/syncthing/syncthing/lib/fs.(*walkFilesystem).walk(0xc420c7d130, 0x4909355, 0x1, 0x49c3a20, 0xc42062a070, 0xc4200aabc0, 0x0, 0xfa631301)
	/Users/jb/src/github.com/syncthing/syncthing/lib/fs/walkfs.go:66 +0x3cb
github.com/syncthing/syncthing/lib/fs.(*walkFilesystem).Walk(0xc420c7d130, 0x4909355, 0x1, 0xc4200aabc0, 0xc4200aabe0, 0xc4200aabd8)
	/Users/jb/src/github.com/syncthing/syncthing/lib/fs/walkfs.go:92 +0x10e
github.com/syncthing/syncthing/lib/scanner.(*walker).fsWalkerRoutine(0xc420676000, 0x49c0c80, 0xc42001ed40, 0xc42008ec60, 0xc4207ea090)
	/Users/jb/src/github.com/syncthing/syncthing/lib/scanner/walk.go:233 +0x106
created by github.com/syncthing/syncthing/lib/scanner.(*walker).walk
	/Users/jb/src/github.com/syncthing/syncthing/lib/scanner/walk.go:122 +0x340
FAIL	github.com/syncthing/syncthing/lib/model	0.134s
```

Happened once and then not again, but looking at the walkFn it's clear our error handling is broken. #4584 removed the early error handing in the walk function.

Of course this becomes clear literally minutes after releasing the RC.

--- a/lib/scanner/walk.go
+++ b/lib/scanner/walk.go
@@ -255,19 +255,15 @@ func (w *walker) createFSWalkFn(ctx context.Context, fsChan chan<- fsWalkResult)
 			skip = fs.SkipDir
 		}
 
-		if path == "." {
-			if err != nil {
-				fsWalkError(ctx, fsChan, path, err)
-				return skip
-			}
-			return nil
-		}
-
 		if err != nil {
 			if sendErr := fsWalkError(ctx, fsChan, path, err); sendErr != nil {
 				return sendErr
 			}
 			return skip
+		}
+
+		if path == "." {
+			return nil
 		}
 
 		if fs.IsTemporary(path) {

--- a/lib/scanner/walk.go
+++ b/lib/scanner/walk.go
@@ -263,6 +263,13 @@ func (w *walker) createFSWalkFn(ctx context.Context, fsChan chan<- fsWalkResult)
 			return nil
 		}
 
+		if err != nil {
+			if sendErr := fsWalkError(ctx, fsChan, path, err); sendErr != nil {
+				return sendErr
+			}
+			return skip
+		}
+
 		if fs.IsTemporary(path) {
 			l.Debugln("temporary:", path)
 			if info.IsRegular() && info.ModTime().Add(w.TempLifetime).Before(now) {
@@ -279,13 +286,6 @@ func (w *walker) createFSWalkFn(ctx context.Context, fsChan chan<- fsWalkResult)
 
 		if w.Matcher.Match(path).IsIgnored() {
 			l.Debugln("skip walking (patterns):", path)
-			return skip
-		}
-
-		if err != nil {
-			if sendErr := fsWalkError(ctx, fsChan, path, err); sendErr != nil {
-				return sendErr
-			}
 			return skip
 		}
 

